### PR TITLE
feat: access-controlled collections, first blood

### PR DESCRIPTION
**Pull request by 3Hren**
This PR introduces a concept of secure storage collections.

By default all collections are considered common, and anyone is allowed to perform read or write operations. But after the first `write` operation a collection is captured by UIDs that performed such operation and becomes owned with `RW` access for UIDs that captured it and with no access to anyone else. This behaviour is handled by recording another secure collection, where all ACL's lay.

For example if an user `42` performed write operation over `secrets` collection and `who-killed-kennedy` key, the following record will be created at `<SECURE>` collection with key `secrets`:

```json
{
    "42": 3
}
```

That means that user `42` has now RW access (0x01 | 0x02) to collection `secrets`. That also means that it can change `<SECURE>` collection itself for keys where there is write access allowed, which It is useful for permission changes.

Moreover this PR adds a new category of logs - audit logs - with strict recording of **who**, **when**, **what** and **where** have been done with collections.

For example:
```
I, 2017-02-15 09:18:55.924: completed 'write' operation [size=22, event=write, collection=secrets, key=key, uids=[42], service=storage, source=audit]

W, 2017-02-15 09:18:55.922: failed to complete 'read' operation [code=2, error=[2] unauthorized, event=read, collection=collection, key=key, uids=[], service=storage, source=audit]
W, 2017-02-15 09:18:55.925: failed to complete 'write' operation [code=13, error=[13] Permission denied, event=write, collection=secrets, key=key, uids=[100500], service=storage, source=audit]
```

Currently all audit logs are written directly into the `core` logger, but this will be changed in further releases.

New component is introduced in this PR: controllers - things that verify access to a service's methods.

To configure `collection` controller, add these lines into the config:

```js
 "controllers": {
    "collection": {
        "type": "collection", // Also possible "null" value, that disables collection control.
        "args": {
            "backend": "core"
        }
    }
}
```

Features done:
- [x] Verbose logging.
- [x] Improved middleware support.
- [x] Audit logs.
- [x] Documentation.
- [x] Config example.

### CODE REVIEW

**Comment by 3Hren** [comment]
@noxiouz @antmat @karitra you can start reviewing process.

File: src/controller/collection.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        switch (event) {
+        case io::event_traits<io::storage::read>::id:
+        case io::event_traits<io::storage::find>::id:
+            return {flag_t::read};
+        case io::event_traits<io::storage::write>::id:
+        case io::event_traits<io::storage::remove>::id:
+            return {flag_t::write};
+        }
+
+        return {flag_t::read};
```
Comment (antmat) [inline src/controller/collection.cpp]: I think it is exceptional situation

File: src/controller/collection.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+namespace defaults {
+
+const std::string collection_acls = ".collection-acls";
+
+} // namespace defaults
+} // namespace
+
+null_t::null_t(context_t& context, const std::string& name, const dynamic_t& args) {
+    (void)context;
```
Comment (antmat) [inline src/controller/collection.cpp]: wtf?

File: src/service/storage.cpp [hunk 3]
```diff
@@ -41,74 +55,213 @@ abort_deferred(cocaine::deferred<T>& def, const std::error_code& ec, const std::
...
+};
+
+} // namespace
 
 storage_t::storage_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args):
     category_type(context, asio, name, args),
     dispatch<storage_tag>(name)
 {
-    const auto storage = api::storage(context, args.as_object().at("backend", "core").as_string());
+    backend = api::storage(context, args.as_object().at("backend", "core").as_string());
```
Comment (antmat) [inline src/service/storage.cpp]: Y U need class member backend?

File: src/service/storage.cpp [hunk 4]
```diff
@@ -41,74 +55,213 @@ abort_deferred(cocaine::deferred<T>& def, const std::error_code& ec, const std::
...
-    on<storage::read>().execute([=](const std::string& collection, const std::string& key) {
-        deferred<std::string> result;
-        storage->read(collection, key, [=](std::future<std::string> future) mutable {
-            try {
-                result.write(future.get());
-            } catch (const std::system_error& e) {
-                abort_deferred(result, e.code(), e.what());
-            }
-        });
+    auto audit = std::shared_ptr<logging::logger_t>(context.log("audit", {{"service", name}}));
```
Comment (antmat) [inline src/service/storage.cpp]: strange name

**Comment by karitra** [comment]
Have read superficially, LGTM

**Pull request merged**